### ISSUE TITLE
Excluder class produces a lot of Strings objects

### ISSUE THREAD

**Issue by olpapchenko** [google/gson#1808]
Hi, 
`com.google.gson.internal.Excluder` class contains method `isAnonymousOrLocal` which in turn calls `clazz.isAnonymousClass()`
```
  private boolean isAnonymousOrLocal(Class<?> clazz) {
    Boolean res = memo.get(clazz);

    if(res != null) {
      return res;
    }

    res = !Enum.class.isAssignableFrom(clazz)
        && (clazz.isAnonymousClass() || clazz.isLocalClass());
    memo.put(clazz, res);
    return res;
  }
```
Inside `isAnonymousClass` performed substring operation - thus each time a new string is created. If the application deserializes/serialize a lot of objects with a lot of fields - we get a lot of strings in the heap that cause frequent GCs
For each class Excluder calls `isAnonymousClass` and each time it creates a substring  - does it make sense to use caching here - so for each particular class only the first time the call to substring is performed?

**Comment by Marcono1234** [comment]
It looks like this is fixed in JDK 11: [JDK-8187123](https://bugs.openjdk.java.net/browse/JDK-8187123)

If this should be solved in Gson, the cache would then probably have to use `WeakReference`s; otherwise it would prevent classes from being unloaded.

----

Note that I am not a maintainer of this project.

**Comment by olpapchenko** [comment]
Thank you @Marcono1234,  Checked on `openjdk version "11.0.7" 2020-04-14 LTS` - the new implementation of `Class.getSimpleName` helped a lot. Strings are not allocated on each call - as the result much less frequent GCs.
